src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.RawSocket.cs

```diff
@@ -0,0 +1,376 @@
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.NetworkInformation
+{
+    public partial class Ping
+    {
+        private const int IcmpHeaderLengthInBytes = 8;
+        private const int MinIpHeaderLengthInBytes = 20;
+        private const int MaxIpHeaderLengthInBytes = 60;
+
+        private unsafe SocketConfig GetSocketConfig(IPAddress address, byte[] buffer, int timeout, PingOptions? options)
+        {
+            // Use a random value as the identifier. This doesn't need to be perfectly random
+            // or very unpredictable, rather just good enough to avoid unexpected conflicts.
+            ushort id = (ushort)Random.Shared.Next(ushort.MaxValue + 1);
+            IpHeader iph = default;
+
+            bool ipv4 = address.AddressFamily == AddressFamily.InterNetwork;
+            bool sendIpHeader = ipv4 && options != null && SendIpHeader;
+
+             if (sendIpHeader)
+             {
+                iph.VersionAndLength = 0x45;
+                // On OSX this strangely must be host byte order.
+                iph.TotalLength = (ushort)(sizeof(IpHeader) + checked(sizeof(IcmpHeader) +  buffer.Length));
+                iph.Protocol = 1; // ICMP
+                iph.Ttl = (byte)options!.Ttl;
+                iph.Flags = (ushort)(options.DontFragment ? 0x4000 : 0);
+#pragma warning disable 618
+                iph.DestinationAddress = (uint)address.Address;
+#pragma warning restore 618
+                // No need to fill in SourceAddress or checksum.
+                // If left blank, kernel will fill it in - at least on OSX.
+             }
+
+            return new SocketConfig(
+                new IPEndPoint(address, 0), timeout, options,
+                ipv4, ipv4 ? ProtocolType.Icmp : ProtocolType.IcmpV6, id,
+                CreateSendMessageBuffer(iph, new IcmpHeader()
+                {
+                    Type = ipv4 ? (byte)IcmpV4MessageType.EchoRequest : (byte)IcmpV6MessageType.EchoRequest,
+                    Identifier = id,
+                }, buffer));
+        }
+
+        private Socket GetRawSocket(SocketConfig socketConfig)
+        {
+            IPEndPoint ep = (IPEndPoint)socketConfig.EndPoint;
+            AddressFamily addrFamily = ep.Address.AddressFamily;
+            SocketType socketType = RawSocketPermissions.CanUseRawSockets(addrFamily) ?
+                SocketType.Raw :
+                SocketType.Dgram; // macOS/iOS has ability to send ICMP echo without RAW
+
+            Socket socket = new Socket(addrFamily, socketType, socketConfig.ProtocolType);
+            socket.ReceiveTimeout = socketConfig.Timeout;
+            socket.SendTimeout = socketConfig.Timeout;
+            if (addrFamily == AddressFamily.InterNetworkV6 && OperatingSystem.IsMacOS())
+            {
+                socket.DualMode = false;
+            }
+
+            if (socketConfig.Options != null && socketConfig.Options.Ttl > 0)
```

Comment:
```suggestion
			if (socketConfig.Options?.Ttl > 0)
```

Reply:
Happy to apply that (and other suggestions). Just note that the code in `Ping.RawSocket.cs` is straight move from `Ping.Unix.cs` with no code changes.